File: backend/python/pipeline/publisher.py
import json
import psycopg2
from psycopg2.extras import RealDictCursor
import redis
from datetime import datetime
from backend.python.config import DB_CONN_STR, REDIS_URL

r = redis.Redis.from_url(REDIS_URL)

def publish_pending_documents():
    """
    Queries Postgres for PENDING staging data and pushes it to Redis.
    """
    total_queued = 0
    queue_name = "company_brain:ingest"

    conn = psycopg2.connect(DB_CONN_STR)
    cursor = conn.cursor(cursor_factory=RealDictCursor)
    
    try:
        #  HARVEST CUSTOM DOCUMENTS
        cursor.execute(
            'SELECT id, "externalId", "rawPayload", "createdAt" FROM "CustomDocument" WHERE status = \'PENDING\' LIMIT 100;'
        )
        custom_docs = cursor.fetchall()
        
        for doc in custom_docs:
            payload = {
                "id": doc["id"],
                "source": "CUSTOM",
                "content": json.dumps(doc["rawPayload"]),
                "metadata": {
                    "title": f"Custom Doc {doc['externalId']}",
                    "timestamp": doc["createdAt"].isoformat() if isinstance(doc["createdAt"], datetime) else str(doc["createdAt"])
                }
            }
            
            # Atomic Pipeline: Push onto Redis queue list

            
            r.lpush(queue_name, json.dumps(payload))
            
            # Update Postgres status to QUEUED
            cursor.execute(
                'UPDATE "CustomDocument" SET status = \'QUEUED\', "queuedAt" = %s WHERE id = %s;',
                (datetime.utcnow(), doc["id"])
            )
            total_queued += 1

        # HARVEST NOTION PAGES
        cursor.execute(
            'SELECT id, title, "plainText", url, "syncedAt" FROM "NotionPage" WHERE status = \'PENDING\' LIMIT 100;'
        )
        notion_pages = cursor.fetchall()
        
        for page in notion_pages:
            payload = {
                "id": page["id"],
                "source": "NOTION",
                "content": page["plainText"] or "",
                "metadata": {
                    "title": page["title"],
                    "url": page["url"],
                    "timestamp": page["syncedAt"].isoformat() if isinstance(page["syncedAt"], datetime) else str(page["syncedAt"])
                }
            }
            
            r.lpush(queue_name, json.dumps(payload))
            cursor.execute(
                'UPDATE "NotionPage" SET status = \'QUEUED\', "queuedAt" = %s WHERE id = %s;',
                (datetime.utcnow(), page["id"])
            )
            total_queued += 1
            
        # Commit changes to database
        conn.commit()
        print(f" Successfully queued {total_queued} documents into Redis.")
        return total_queued
        
    except Exception as e:
        conn.rollback()
        print(f" Error during pipeline flush: {e}")
        raise e
    finally:
        cursor.close()
        conn.close()

File: lib/queue/publisher.ts
import Redis from 'ioredis';
import prisma from '../prisma';

const redisUrl = process.env.REDIS_URL || 'redis://localhost:6379/0';
export const redis = new Redis(redisUrl);

// Uniform payload contract expected by Python/Celery worker
interface QueuePayload {
  id: string;
  source: 'SLACK' | 'NOTION' | 'GMAIL' | 'GITHUB' | 'CUSTOM';
  content: string;
  metadata: {
    title?: string;
    author?: string;
    url?: string;
    timestamp?: string;
  };
}

export async function publishPendingDocuments(): Promise<{ queuedCount: number }> {
  let totalQueued = 0;
  const targetStatus = 'PENDING';
  const queuedStatus = 'QUEUED';
  const queueName = 'company_brain:ingest'; // Match Redis dataflow queue name

  // --- 1. PROCESS CUSTOM DOCUMENTS ---
  const pendingCustomDocs = await prisma.customDocument.findMany({
    where: { status: targetStatus },
    take: 50, // Batch limit to preserve memory
  });

  for (const doc of pendingCustomDocs) {
    const payload: QueuePayload = {
      id: doc.id,
      source: 'CUSTOM',
      content: JSON.stringify(doc.rawPayload),
      metadata: {
        title: `Custom Doc ${doc.externalId}`,
        timestamp: doc.createdAt.toISOString(),
      },
    };

    // Atomic transaction: Push to Redis + Update Postgres Status
    await redis.lpush(queueName, JSON.stringify(payload));
    await prisma.customDocument.update({
      where: { id: doc.id },
      data: { status: queuedStatus, queuedAt: new Date() },
    });
    totalQueued++;
  }

  // --- 2. PROCESS NOTION PAGES ---
  const pendingNotionPages = await prisma.notionPage.findMany({
    where: { status: targetStatus },
    take: 50,
  });

  for (const page of pendingNotionPages) {
    const payload: QueuePayload = {
      id: page.id,
      source: 'NOTION',
      content: page.plainText || '',
      metadata: {
        title: page.title || undefined,
        author: page.createdByName || undefined,
        url: page.url || undefined,
        timestamp: page.syncedAt.toISOString(),
      },
    };

    await redis.lpush(queueName, JSON.stringify(payload));
    await prisma.notionPage.update({
      where: { id: page.id },
      data: { status: queuedStatus, queuedAt: new Date() },
    });
    totalQueued++;
  }

  return { queuedCount: totalQueued };
}